### ISSUE TITLE
POST Request with cURL client not working

### ISSUE THREAD

**Issue by collizo4sky** [hybridauth/hybridauth#814]
Making POST request with `return $this->apiRequest($url, 'POST', $body)` not working.

Note: am trying to post a JSON to the API endpoint but the API is returning invalid JSON.

I feel the error is from here https://github.com/hybridauth/hybridauth/blob/master/src/HttpClient/Curl.php#L119

FWIW: see possible solution https://lornajane.net/posts/2011/posting-json-data-with-php-curl

I actually fix the problem replacing the line with below and with a `application/json` header.

`$this->curlOptions[CURLOPT_POSTFIELDS] = json_encode($parameters);`

**Comment by miled** [comment]
Because certain apis may not accept json as content type, one solution is to change the current implementation of apiRequest to make it work with either an associative array of parameters or a pre formatted string as payload, ex:

```php
$adpter->apiRequest($url, 'POST',
  json_encode(...),
  ['Content-Type' => 'application/json', 'Content-Length' => ...]);
```

**Comment by collizo4sky** [comment]
@miled I am using this with ConstactContact API and they require payload must be JSON.

We could add a check/if-condition to detect if the header contains `application/json` content type then run json_encode() on the payload.

What do you think?

I could whip up a PR for both curl and guzzle client if this proposal is accepted.

**Comment by miled** [comment]
Sounds fairly reasonable, go for it.

>PUT and DELETE http verb support is missing

They weren't implemented simply because I haven't come across a use case that requires them, but there are no reasons not to support those as well.

**Comment by collizo4sky** [comment]
Fixed this issue in #816 and #817

**Comment by miled** [comment]
Great. Thanks.
